src/main/java/com/iota/iri/MilestoneTracker.java

```diff
@@ -0,0 +1,549 @@
+        if(targetMilestone == null || targetMilestone.index() < initialSnapshot.index()) {
+            return;
+        }
+
+        // block write access to the ledger from other threads
+        lock.writeLock().lock();
+
+        // create a progress logger and start the logging
+        ProgressLogger hardResetLogger = new ProgressLogger(
+            "Resetting ledger to milestone " + targetMilestone.index() + " due to \"" + reason + "\"", log
+        ).start(latestMilestoneIndex - targetMilestone.index() + 2); // +1 for softReset and +1 for starting milestone
+
+        // prune all potentially invalid database fields
+        try {
+            MilestoneViewModel currentMilestone = targetMilestone;
+            while(currentMilestone != null) {
+                // reset the snapshotIndex of the milestone and its referenced approvees + it's StateDiff
+                resetSnapshotIndexOfMilestone(currentMilestone);
+                tangle.delete(StateDiff.class, currentMilestone.getHash());
+
+                int currentStep = latestMilestoneIndex - currentMilestone.index() + 2 - hardResetLogger.getStepCount();
+                hardResetLogger.progress(currentStep);
+
+                // iterate to the next milestone
+                currentMilestone = MilestoneViewModel.findClosestNextMilestone(tangle, currentMilestone.index());
+            }
+        } catch(Exception e) {
+            hardResetLogger.abort(e);
+        }
+
+        // after we have cleaned up the database we do a soft reset to rescan the existing data
+        softReset();
+
+        // dump message when we are done
+        hardResetLogger.finish();
+
+        // unblock write access to the ledger from other threads
+        lock.writeLock().unlock();
+    }
+
+    /**
+     * This method resets the snapshotIndex of all transactions that "belong" to a milestone.
+     *
+     * While traversing the graph we use the snapshotIndex value to check if it still belongs to the given milestone and
+     * ignore all transactions that where referenced by a previous one. Since we check if the snapshotIndex is bigger or
+     * equal to the one of the targetMilestone, we can ignore the case that a previous milestone was not processed, yet
+     * since it's milestoneIndex would still be 0.
+     *
+     * @param currentMilestone the milestone that shall have its confirmed transactions reset
+     * @throws Exception if something goes wrong while accessing the database
+     */
+    public void resetSnapshotIndexOfMilestone(MilestoneViewModel currentMilestone) throws Exception {
+        // initialize variables used for traversing the graph
+        TransactionViewModel milestoneTransaction = TransactionViewModel.fromHash(tangle, currentMilestone.getHash());
+        Set<Hash> seenMilestoneTransactions = new HashSet<>();
+        Queue<TransactionViewModel> transactionsToExamine = new LinkedList<>(Collections.singleton(milestoneTransaction));
```

Comment:
ArrayLists are usually faster, so probably use `ArrayDeque`. 
If you can estimate its size (doesn't have to be exact) you can improve performance.